news/2016-05-18-enhanced-instance.asciidoc

```diff
@@ -0,0 +1,174 @@
+priority: 1
+change_frequency: weekly
+excerpt: Weld team is considering adding an enhanced version of javax.enterprise.inject.Instance to the Weld API. Explore the possibilities...
+desc: Experiments with enhanced version of javax.enterprise.inject.Instance
+---
+:linkattrs:
+
+`javax.enterprise.inject.Instance` is sometimes invaluable companion.
+Not only it allows to handle unsatisfied dependency gracefully, i.e. to avoid aborting container initialization unless really necessary:
+
+[source,java]
+----
+interface OrderProcessor {
+  void process(Order order);
+  int getPriority();
+}
+
+class OrderService {
+
+  @Inject
+  Instance<OrderProcessor> instance;
+
+  void create(Order order) {
+    if (!instance.isUnsatisfied() && !instance.isAmbiguous()) {
+        instance.get().process(order);
+    } else {
+        // Log a warning or throw an exception
+    }
+  }
+}
+----
+
+It is less known that `Instance` extends `Iterable` and so it allows to iterate contextual references of beans with the specified combination of required type and qualifiers.
+This might be useful if the set of beans satisfying the given type and qualifiers is not known beforehand.
+We may also need to resolve ambiguities manually, e.g. inspect all the instances and choose only those matching our needs.
+
+[source,java]
+----
+@ApplicationScoped
+class OrderService {
+
+  @Inject
+  @Any
+  Instance<OrderProcessor> instance;
+
+  void create(Order order) {
+    for (OrderProcessor processor : instance) {
+        if (processor.getPriority() > 10) {
+            processor.process(order);
+        }
+    }
+  }
+}
+----
+
+Another interesting use case might be selecting exactly one implementation:
+
+[source,java]
+----
+@ApplicationScoped
+class OrderService {
+
+  @Inject
+  @Any
+  Instance<OrderProcessor> instance;
+
+  void create(Order order) {
+    List<OrderProcessor> processors = new ArrayList<>();
+    for (OrderProcessor processor : instance) {
+        processors.add(processor);
+    }
+    Collections.sort(processors, Comparator.<OrderProcessor> comparingInt(p -> p.getPriority()).reversed());
+    // Use the processor with highest priority
+    processors.get(0).ping();
+  }
+}
+----
+
+This works nice.
+But we have to pay extra attention to the scopes.
+If a `Processor` implementation is `@Dependent` a similar usage results in memory leaks.
+What's the reason?
+The `Processor` instance is the dependent object of `Instance<OrderProcessor>` which is the dependent object of `OrderService`.
+And so each `Instance.get()` will produce a new `Processor` bound to the lifecycle of the `OrderService`.
+To avoid the leak we should always call `Instance.destroy()` method in similar cases.
+However, `Instance.destroy()` always destroys the underlying contextual instance!
+Even if you pass e.g. a client proxy of an `@ApplicationScoped` bean.
+This is not always desirable.
+Currently, it would be quite complicated to detect a dependent bean inside the loop.
```

Comment:
`it would be quite` -> `it is quite`
Present simple because it just is complicated and that is why WeldInstance is awesome, right? :)